### DISCUSSION TITLE
Adding HTTP method for ECU_R_PRO and ECU_C inverters

### DISCUSSION THREAD

**Discussion by ksheumaker** [ksheumaker/homeassistant-apsystems_ecur#46] [Ideas]
Talk about adding a new query method via the internal web interface

**Comment by ksheumaker** [comment]
@tv3 - We can use this to discuss adding the http query method you are working on.

The best way forward I see is to create two classes, change `APSystemsECUR` to `APSystemsSocket` and then yours can be `APSsystemsHTTP`.  If both classes have the same method (i.e. query_ecu) and return the same data structure.  I can then use them interchangeably in `__init__.py`. I. would create the correct object depending on what config option you select when add the integration.  I figured i'd add a drop down to pick socket or HTTP, with a description to tell them what they should use depending on the type of ECU.

I see you do still query the ECU via socket interface once on line 237.  What data is not available in the webpage you need to get from the socket method?  I wonder if it makes sense to alter the data returned from the integration to just skip that un-available data if you use the HTTP method.

**Comment by tv3** [comment]
I've uploaded some example files in my fork. Screenshots and some curl output.

repo: https://github.com/tv3/homeassistant-apsystems_ecur

power_graph, energy_graph and realtimedata are the actual webpages with calls to old_ urls to call the webapis.
example: curl -X POST http://192.168.4.6/index.php/realtimedata/old_energy_graph -d "date=2022-02-17"

example files: https://github.com/tv3/homeassistant-apsystems_ecur/tree/main/files

realtimedata=inverter data
old_power_graph=generated power of specific day (like in EMA app)
old_energy_graph=weekly/monthly/yearly generated power data (like in EMA app)

realtimedata doesn't have a visible webapi, so I created a webpagwe scraper to harvest the data, but it supplies all the inverter.

As for the TCP query, I use:
I haven't found basic info on the ECU or the inverters in the webpages, so I grab those initially through the TCP query. I only need those once and are stored in the ECU object.
The TCP method can be used, but only sparingly. I think there's some kind of memory leak or other bug that breaks the data structures in the ECU. I've seen missing data in the realtimedata webpage when the TCP queries were getting instable.
My initial approach was to mimick the response that your integration generates to make it a drop-in for the async query.

Splitting would make development a bit easier, as the code can be developed as separate modules.

Now how to go forward. My fork was just for getting code to work. How should we merge my code into your repo?

**Comment by ksheumaker** [comment]
Going through the code here's the things it looks like you haven't found in the http method:

- ecu_id
- lifetime_energy
- timezone
- firmware

and on the inverter side looks like we are missing:

- model
- signal

Is the ecu id, firmware, or timezone available under the administration tab by any chance?

I kind of think missing some of these are ok, if we can stay 100% http.  It seems best to not mix these if possible.  I can certainly make the integration provide slightly different entities depending on query method.

I'm going to publish a new version of the code today that moves APSystemsECUR into APSystemsSocket and has some commented code in the config flow that will produce this:

<img width="672" alt="Screen Shot 2022-02-19 at 11 05 45 AM" src="https://user-images.githubusercontent.com/12651/154811849-e880bd6a-dc9c-41f0-83be-e69666b07c05.png">

If you want to submit a merge request of a new class `APSystemsHTTP` that handles the HTTP requests that would be great.  For stability reasons on the ECU this latest version has moved away from asyncio socket code, so if you want to use something like the python `requests` library which I see you imported that's fine.  I'm letting HA handle running our non async integration in an async context by using their `hass.async_add_executor_job` method.

If you don't want to work on the HTTP class I can do it too, will just be a bit harder for me without the appropriate hardware.

**Reply by tv3** [reply]
Haven't found that data yet in the admin pages. 
I'm thinking about the missing parts (static values like ecu/inverters). They are relevant for the entity discovery and registry in HA, right?
If I could get the values from those entities, I wouldn't need the socket call at all. Will work to kill off the tcp call. 

Do you have info on querying entities in HA?

Additional step will be investigating service mode and see if that reveals more data.

**Reply by ksheumaker** [reply]
Currently the entities are made with the ECU ID and Inverter ID, but we could fix that if we don't have those values via HTTP.  It does look like you can grab the inverter ID but not the ECU.  

This is the code that generates the device registry is in `__init__.py`.  Things like model, and sw_version are optional.  And the identifier just need to be unique to the integration, so we could not use `ecu_<ecu_id>` but just `ecu` since there is only one.   I can certainly just change the name to "ECU" instead of "ECU XXXXX" as well.  Should be pretty trivial to adapt to whatever data you can get. 

```
    device_registry = dr.async_get(hass)

    device_registry.async_get_or_create(
        config_entry_id=config.entry_id,
        identifiers={(DOMAIN, f"ecu_{ecu.ecu.ecu_id}")},
        manufacturer="APSystems",
        suggested_area="Roof",
        name=f"ECU {ecu.ecu.ecu_id}",
        model=ecu.ecu.firmware,
        sw_version=ecu.ecu.firmware,
    )

    inverters = coordinator.data.get("inverters", {})
    for uid,inv_data in inverters.items():
        model = inv_data.get("model", "Inverter")
        device_registry.async_get_or_create(
            config_entry_id=config.entry_id,
            identifiers={(DOMAIN, f"inverter_{uid}")},
            manufacturer="APSystems",
            suggested_area="Roof",
            name=f"Inverter {uid}",
            model=inv_data.get("model")
        )
```

To me the advantages of never doing the socket method on the http version outweigh a few of the missing pieces of data, that aren't necessary to have.

**Comment by ksheumaker** [comment]
v1.2.11 has been release and is now the main branch of the repo.

**Comment by tv3** [comment]
Service mode doesn't add anything to supply additional info. So that was a dead end. 

I did manage to scrape the ecu_id and have added it to the http_query. Nice to have some unique id in the resultset. The model attribute is already known when setting up the integration, so there is no need for that in the active querying, imho.
The signal data is't available, but I think we're not missing out on much intergation wise.

added some additional data: peak/mean/median power current day and for the last 7 days (week). 

Now the http_query update module is running without the tcp functions. Need to further cleanup etcetera and then I'll update my repo.

Next step is how to split modules & integrate it into your repo.

updated my repo: https://github.com/tv3/homeassistant-apsystems_ecur/blob/main/custom_components/apsystems_ecur/APSystemsECUR.py

**Comment by ksheumaker** [comment]
I'll have some time this weekend, and I can integrate it in if you'd like me to do that?

**Reply by tv3** [reply]
I think that would be the best way forward. You could first merge 'as-is'. 
The update method could be dynamic in the future. When http Port is available, use http query... just thinking out load ;-)

**Comment by tv3** [comment]
@ksheumaker I've seen people using my http version with succes. But updating to 1.2.x breaks it. I checked the versions and saw that you moved away from async IO. So I'm adjusting my code to be sync instead of async and putting it in APSystemsHTTP.py
Effectively I'm separating the http code from the socket code as you had already done in 1.2.x.

Created a stripped down APSsytemsHTTP.py version. This could potentially be easily merged with your current 1.2.x release. I need to check all the things that have changed (like the query switch etc) and see how that impacts my 1.1 setup. 

https://github.com/tv3/homeassistant-apsystems_ecur/blob/main/custom_components/apsystems_ecur/APSystemsHTTP.py

**Comment by tv3** [comment]
@ksheumaker 
Been working this weekend on a version that would serve both HTTP and TCP integrations as there were questions from users concerning the ecu-r-pro stability.
So I've started a new branch in my repo: https://github.com/tv3/homeassistant-apsystems_ecur/tree/tv3-standalone-http-integration
It is still based on your 1.1.2 integration, but I picked out some changes (like to IO stuff) out of your main branch in 1.2. Changed to sync IO too.
I've implemented a kind of modular approach by putting your TCP code in a module and my HTTP code in a module and keeping the APSystemsECUR as a wrapper. Didn't want to change all the other HA stuff, as I don't really know the HA details.
